### PR TITLE
Fix test on gpu

### LINKED ISSUE #2459: GradsScalarHandler logs 0 gradients if default update function is used
## 🐛 Bug description

Logging the gradients per epoch / iteration is a useful way to debug an under-performing model. Ignite provides an easy-to-use tensorboard_logger handler, an example accessible from `ignite.contrib.handlers.tensorboard_logger.GradsScalarHandler`. However, the default `update` function used by Engines generated by `create_supervised_trainer` zero the gradients before terminating, causing the handler to log zeroed out gradients all the time. 

**Steps to reproduce:**
My code is too complicated at the momemt to provide a clear insight, and I am limited by time to provide a minimal (not-)working example, so I will provide abstracted steps.

1. Generate an `Engine` / `DeterministicEngine` on an arbitrary problem by the `create_supervised_trainer` method.
2. Establish a `TensorboardLogger` and a .
3. Attach a `GradsScalarHandler` / your choice of a gradient logger. Also log the training loss or some other metric.
4. Start the training, check tensorboard and see the constant-0 gradient norms / gradients, despite the losses/metrics implying some sort of improvement/learning takes place.

**Solution proposal**

Preserving the gradients until epoch end is tricky, but not required for my purposes. If we are OK with using  `Events.ITERATION_COMPLETED` as a cue to log gradients, then we can simply modify the default `update` functions as follows:

(assuming  `engine.state.iteration` counts from 1).
```
    def update(engine: Engine, batch: Sequence[torch.Tensor]) -> Union[Any, Tuple[torch.Tensor]]:
        if (engine.state.iteration + 1) % gradient_accumulation_steps == 0:
            optimizer.zero_grad()
        model.train()
        x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
        y_pred = model(x)
        loss = loss_fn(y_pred, y)
        if gradient_accumulation_steps > 1:
            loss = loss / gradient_accumulation_steps
        loss.backward()
        if engine.state.iteration % gradient_accumulation_steps == 0:
            optimizer.step()
        return output_transform(x, y, y_pred, loss)
```

This way, upon completion of `update` and at the moment of `Events.ITERATION_COMPLETED` firing, there will be some non-zero gradients available to be logged.

## Environment (latest version of Ignite still has the same bug)

- PyTorch Version (e.g., 1.4): 1.10.1
- Ignite Version (e.g., 0.3.0): 0.4.7
- OS (e.g., Linux):
- How you installed Ignite (`conda`, `pip`, source): conda
- Python version: 3.9.7
- Any other relevant information:

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Fixes #2459 

Description:
There was an issue in `test_create_supervised` tests causing device mismatch error.

### CODE REVIEW

File: tests/ignite/engine/test_create_supervised.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def _():
         assert pytest.approx(trainer.state.output[-1], abs=1e-5) == loss[0]
         accumulation[0] = loss[0] = 0.0
 
-    if model_device == trainer_device or ((model_device == "cpu") ^ (trainer_device == "cpu")):
+    if model_device == trainer_device or ((model_device in ["cpu", None]) and (trainer_device in ["cpu", None])):
```
Comment (sadra-barikbin) [inline tests/ignite/engine/test_create_supervised.py:109]: I changed this because in the case `trainer_device` and `model_device` were `cuda` and `cpu` or vice versa, we would not catch device mismatch error.

File: tests/ignite/engine/test_create_supervised.py [hunk 2]
```diff
@@ -124,6 +124,40 @@ def _():
...
+
+    x = torch.tensor([[1.0], [1.0], [1.0], [1.0], [1.0]])
+    y = torch.tensor([[2.0], [3.0], [4.0], [5.0], [6.0]])
+    data = [(_x, _y) for _x, _y in zip(x, y)]
+
+    @trainer.on(Events.ITERATION_COMPLETED)
+    def _():
+        assert model.weight.grad != 0
+
+    if model_device == trainer_device or ((model_device in ["cpu", None]) and (trainer_device in ["cpu", None])):
```
Comment (sadra-barikbin) [inline tests/ignite/engine/test_create_supervised.py:153]: I copied this statement from `_test_create_supervised_trainer`.

**Review by vfdev-5 — APPROVED** [review]
Thanks @sadra-barikbin
